### ISSUE TITLE
Redmine not seeing run_git_as_git_user

### ISSUE THREAD

**Issue by garg** [ericpaulbishop/redmine_git_hosting#53]
I am running CentOS 5.6, with redmine 1.2 installed and running. When I try to add a git repository from within a redmine project, I see:

"Scm command is not available. Please check settings on the administration panel." 

When I go to the administration panel>Settings>Repositories, I see:

[x] Git (!)  /tmp/redmine_git_hosting/run_git_as_git_user

I checked and the file   /tmp/redmine_git_hosting/run_git_as_git_user exists and is accessible by git. and apache

When I check visudo I have:

apache  ALL=(git)       NOPASSWD:ALL
git     ALL=(apache)    NOPASSWD:ALL

When I look the redmine_git_hosting settings page I see 

Git user is able to sudo to Web user? (!)
Web user is able to sudo to Git user?  (!) 

Any idea why redmine can't see /tmp/redmine_git_hosting/

Thanks

**Comment by hcartiaux** [comment]
Maybe a hint for you. I got a similar problem today. I use mod_passenger for Apache, so the default user of redmine is nobody, and it did not match with my sudoers file.
Adding "PassengerDefaultUser www-data" to my Apache configuration has fixed the problem ;)

**Comment by allen13** [comment]
I am also having the exact same problem as garg, additionally I am running the same setup as well. I have tried hcartiaux's advice to no avail. 

I went into redmine's git_adapter.rb and manually set the version so it would pass the scm check. Unfortunately the scm check was doing its job properly, so the first time the system executed a git command it threw an error page.

I am running apache with virtualmin suexec so the redmine project gets run as the local virtualhost user. When I run run_git_as_git_user as the vh user it fails because git is not in the path. I changed the sudo configuration to keep the PATH variable. This allowed the script to run from the command line but still no change on the server side.

I'm pretty sure the problem has something to do with the PATH that the mod_passenger server is seeing but I don't know how fix this.

**Comment by ericpaulbishop** [comment]
Yeah... the "web server" user should be the one you're running Rails under, and that user needs to have the git executable in its PATH.  I've added a note about the "web server user" being the Rails user, not necessarily the main web server user in the instructions for the latest versions.  There is no good way around requiring the git executable in the PATH of the web user (and the git user).

**Comment by garg** [comment]
Unfortunately, I am still getting exclamation marks in two places.
1. <redmine>/settings/plugin/redmine_git_hosting, it shows Git user is able to sudo to Web user? and Web user is able to sudo to Git user? as exclamation marks. This happens even though I can confirm in the terminal that  users git and sudo work both ways.
2. <redmin>/settings -> repositories shows that git has an exclamation mark infront of /tmp/redmine_git_hosting/run_git_as_git_user how ever when i log on as user apache, i can run that file just fine.

Here is what I have:

Centos 5.6 running redmine 1.2.1 using passenger as user apache

```
# ps aux | grep redmine

apache   20088  9.5 10.8 272152 114024 ?       S    13:01   0:10 Passenger ApplicationSpawner: /usr/local/redmine
apache   20103  1.4 10.9 276648 114528 ?       S    13:01   0:01 Rails: /usr/local/redmine
apache   20076  0.0  0.5 242256  6016 ?        S    13:01   0:00 /usr/sbin/httpd
```

If I switch to user apache this is what I see:

```
[root@redmine redmine]# su apache

bash-3.2$ echo $PATH
/usr/kerberos/sbin:/usr/kerberos/bin:/usr/local/sbin:/usr/local/bin:/sbin:/bin:/usr/sbin:/usr/bin:/root/bin

bash-3.2$ /tmp/redmine_git_hosting/run_git_as_git_user
usage: git [--version] [--exec-path[=<path>]] [--html-path] [--man-path] [--info-path]

bash-3.2$ sudo -u git whoami
git

bash-3.2$ sudo -u git env
[...]
USER=git
USERNAME=git
PATH=/usr/bin:/bin 
SUDO_COMMAND=/bin/env
[...]
```

The same if I switch user to git. I am able to run things as user apache. 

I have also changed passengerruby to the ruby187env as shown in the install docs.

I then tried reinstalling redmine, passenger, and this plugin from scratch on a different server, and got the same thing.

However, if I run redmine as user 'git' then I have check marks and it sees everything. 

Has anyone run into this?

Thanks :)

**Comment by garg** [comment]
Also, just for debugging I put in: 

```
redmine-2 www # cat redmine/app/views/admin/index.rhtml
<%=h (%x[whoami]).chomp.strip %>
<%=h (%x[env]).chomp.strip %>
```

and I see

```
apache 
TERM=screen PATH=/usr/local/ruby1.8.7/lib/ruby/gems/1.8/gems/passenger-3.0.9/bin/:/usr/local/ruby1.8.7/lib/ruby/gems/1.8/gems/rails-2.3.11/bin:/sbin:/bin:/usr/sbin:/usr/bin:/usr/local/sbin:/usr/local/bin     PWD=/ LANG=C SHLVL=3 _=/usr/local/bin/ruby USER=apache HOME=/var/www RACK_ENV=production RAILS_ENV=production     SERVER_SOFTWARE=Apache/2.2.15 (CentOS) 
```

on the admin page.

**Comment by garg** [comment]
debug logs show 

scm: error during get scm available: Permission denied - /tmp/redmine_git_hosting/gitolite_admin_ssh

Seems like I have a problem with the /var/www/redmine/.ssh/git_gitolite_admin_id_rsa.pub ?

**Comment by garg** [comment]
This is very strange to me:

Here is the code from lib/git_hosting.rb with extra logging added inside

```
test = %x[#{GitHosting.git_user_runner} sudo -nu #{web_user} echo "yes" ]
logger.info GitHosting.git_user_runner
logger.info "test #{test}"
```

the logs then show:

/tmp/redmine_git_hosting/run_as_git_user
test

Basically test never gets populated.

If I change it to:
    test = `whoami`  #backticks 

the logs show
test apache 

if I log in as apache on the CLI and type I get the following:

```
bash-4.1$ /tmp/redmine_git_hosting/run_as_git_user sudo -nu apache echo "yes"
yes
```

All the visudo info is okay but the test variable never gets 'yes' even though running it manually works fine?

**Comment by ghost** [comment]
Did you try running that test as the user that your Redmine installation is running under (www-data for apache)?

**Comment by garg** [comment]
Yes, I ran it as apache but no luck. There seems to be some misconfiguration somewhere..

**Comment by ericpaulbishop** [comment]
First, sorry it's taken a while to get back to you.  I have a lot of stuff going on at once, and this project is often not my first priority.

Second, just to rule out an issue with PATH altogether, put this line in cat redmine/app/views/admin/index.rhtml:

```
<%=h (%x[sudo -V]).chomp.strip %>
```

This will verify you can actually run sudo (if it prints the version info).  If it doesn't, then it is a PATH issue.  Let me know the result, and we'll proceed from there.

**Comment by garg** [comment]
Thank you so much for helping. 

Here is what I have in the code:

<%=h (%x[whoami]).chomp.strip %>
<%=h (%x[env]).chomp.strip %>

<%=h (%x[sudo -V]).chomp.strip %>

and the output is:

apache 

TERM=xterm PATH=/usr/local/ruby1.8.7/lib/ruby/gems/1.8/gems/passenger-3.0.9/bin/:/usr/local/ruby1.8.7/lib/ruby/gems/1.8/gems/rails-2.3.11/bin:/sbin:/bin:/usr/sbin:/usr/bin:/usr/local/sbin:/usr/local/bin:/tmp:/:/tmp/redmine_git_hosting PWD=/ LANG=C SHLVL=3 _=/usr/local/bin/ruby USER=apache HOME=/var/www RACK_ENV=production RAILS_ENV=production SERVER_SOFTWARE=Apache/2.2.15 (CentOS) 

Sudo version 1.7.2p2

**Comment by sboisson** [comment]
I have the same problem (exclamation mark and no access to repository) with a different config (Chiliproject/2.2.0 - Passenger/3.0.9 - nginx/1.0.6 - CentOS/5.6-x86_64)

The gitolite was already install and worked very well prior the plugin install.
git/1.7.3.4

When the sudo test is run, the context is:

<pre>whoami: "projects" 
sudo version: "Sudo version 1.7.2p1" 
PATH=/usr/local/rvm/gems/ree-1.8.7-2011.03/bin:/usr/local/rvm/gems/ree-1.8.7-2011.03@global/bin:/usr/local/rvm/rubies/ree-1.8.7-2011.03/bin:/bin:/usr/bin
</pre>


The sudo config works, when testing from shell:
As user 'projects', running <code>sudo -nu git echo "yes"</code> display <code>yes</code>
As user 'git', running <code>sudo -nu projects echo "yes"</code> display <code>yes</code>

In my case, the sudo seems to work, as whenever I go to the plugin config pages, I can read in the ChiliProject logs the following :

<pre>
Testing if web user("projects") can sudo to git user("git")
"post-receive" not handled by gitolite, installing it...
Installing "post-receive.redmine_gitolite.rb" from /vol/data/projects/chiliproject/vendor/plugins/redmine_git_hosting/contrib/hooks/post-receive.redmine_gitolite.rb to ~/.gitolite/hooks/common/post-receive
"post-receive.redmine_gitolite installed
Running "gl-setup" on the gitolite install...
Finished installing hooks in the gitolite install...
</pre>

**Comment by ericpaulbishop** [comment]
Add the following debug line to the ones above:

<%=h (%x[ sudo -nu #{Setting.plugin_redmine_git_hosting['gitUser'] } echo "yes"  2>&1]).chomp.strip %>

That should add any errors to the output of that command, so we can see exactly what's going wrong.

**Comment by sboisson** [comment]
Thanks Eric,

Here is what I got: <code>sudo: sorry, you must have a tty to run sudo</code>

**Comment by ericpaulbishop** [comment]
Bingo.  This explains it -- the requiretty directive is set for sudo.  Hadn't thought about that, but that can definitely cause problems (I should update the readme).  The best solution is to add the following lines to your sudoers file:

```
Defaults:[web user] !requiretty
Defaults:[git user] !requiretty
```

Note that these lines should go BELOW the existing default directive where tty is set in the first place.

That should eliminate the requiretty directive for the necessary users, while keeping it in place as a security restriction for the others.

**Comment by garg** [comment]
Same here :) 

```
sudo: sorry, you must have a tty to run sudo
```

Fixed! Thanks

**Comment by sboisson** [comment]
Added the two default lines in sudoer file, and I get green checkmark in plugin admin page :-)

Log now:

<pre>
Testing if web user("projects") can sudo to git user("git")
Creating MD5 digests for Redmine Git Hosting hook
Digest for post-receive.redmine_gitolite.rb: 08e9b3cf99fa6285331ccdcef8b9a0d3
Our hook is already installed
</pre>


I still have no access to repositories (the plugin thinks the repositories are empty)

**Comment by ericpaulbishop** [comment]
Have you tried restarting the web server?  A lot of stuff gets cached, and with the original error you may need to make sure the cache gets cleared.

**Comment by ericpaulbishop** [comment]
Any chance you tried setting an ssh key with a password set as the private key to access gitolite?  That's what some guy tried yesterday, and needless to say, it didn't work (and he got this error).  It indicates you're not able to clone the gitolite-admin repository properly.

I'll suggest the same debugging technique to you as I did to him:

Try this: Edit the git_hosting.rb file in [chiliproject_root]/vendor/plugins/redmine_git_hosting/lib/git_hosting.rb, at line 273
Before editing it should look like:

```
%x[env GIT_SSH=#{gitolite_ssh()} git clone #{git_user}@#{Setting.plugin_redmine_git_hosting['gitServer']}:gitolite-admin.git #{local_dir}/gitolite-admin]
```

Update so that it dumps both stdout and stderr from this command to a file in /tmp, so it now reads

```
%x[env GIT_SSH=#{gitolite_ssh()} git clone #{git_user}@#{Setting.plugin_redmine_git_hosting['gitServer']}:gitolite-admin.git #{local_dir}/gitolite-admin >/tmp/CLONE_TEST.txt 2>&1]
```

Then remove what is currently in /tmp/redmine_git_hosting and restart the web server.  Wait until it fails again and post the contents of /tmp/CLONE_TEST.txt here

**Comment by sboisson** [comment]
I found what was the problem with the repository not showing...

Before I installed this plugin, I was using the standard GIT support and the existing git module on gitolite was using a different name than the project id.
After I installed the plugin, ChiliProject was trying to access a GIT module named with the project id..

So as a workaround I created a new ChiliProject project with a project id = git module name.
And I can access the repository..

Regarding the cloning of the gitolite admin, it was my mistake.. the path to the key was wrong.. 
Maybe it would be a good idea to add a check mark to indicate the admin key are found.. :-)

**Comment by sboisson** [comment]
And by the way, you're right about having a password on the private key to access gitolite:

<pre>
Cloning into /tmp/redmine_git_hosting/gitolite-admin...
Permission denied (publickey,gssapi-with-mic).
fatal: The remote end hung up unexpectedly
</pre>

**Comment by isaacavery** [comment]
Hey guys -

I recently ran into this issue (or a version of it at least) when installing Redmine using Apache/Passenger on a debian install. I was able to resolve this issue in two steps:
1. (kudos to hcartiaux above for this solution) Add 'PassengerDefaultUser www-data' to the vhost config file
2. Update the git path in [REDMINE_PATH]/config/configuration.yml (using the line "scm_git_command: /usr/lib/git-core/git")

NOTE: I determined the path to the executable using the command "git exec-path" which is the DIRECTORY of the executable (in my case "/usr/lib/git-core"). Make sure to check that dir and confirm that the exec is there, then add it to the exec-path as "/git"

Hopefully that helps someone out there with the issues that I ran accross.
